tools/lib/NQP/Config.pm

```diff
@@ -0,0 +1,1134 @@
+}
+
+sub nfp {
+    my ( $vol, $dirs, $file ) = File::Spec->splitpath(shift);
+    return File::Spec->catpath( $vol,
+        File::Spec->catdir( File::Spec->splitdir($dirs) ), $file );
+}
+
+sub system_or_die {
+    my @cmd = @_;
+    system(@cmd) == 0
+      or die "Command failed (status $?): @cmd\n";
+}
+
+sub parse_revision {
+    my $rev       = shift;
+    my $sep       = qr/[_.]/;
+    my $rev_regex = qr/
+        (?<year> \d+)
+        $sep
+        (?<month> \d+)
+        (?:
+            $sep
+            (?<day> \d+)
+        )?
+        (?:
+            -
+            (?:
+                (?<revno> \d+) - g[a-f0-9]*
+
+                |
+
+                RC (?<rcno> \d+)
+            )
+        )?
+        $
+    /x;
+    if ( $rev =~ $rev_regex ) {
+        return ( $+{year}, $+{month}, $+{day} // 0, $+{rcno} // 0,
+            $+{revno} // 0 );
+    }
+    else {
+        die "Unrecognized revision specifier '$rev'\n";
+    }
+}
+
+sub cmp_rev {
+    my ( $a, $b ) = @_;
+    my @a   = parse_revision($a);
+    my @b   = parse_revision($b);
+    my $cmp = 0;
+    for ( 0 .. 4 ) {
+        $cmp = $a[$_] <=> $b[$_] if ( defined $a[$_] && defined $b[$_] );
+        last if $cmp;
+    }
+    $cmp;
+}
+
+sub read_config {
+    my @config_src = @_;
+    my %config     = ();
+    local $_;
+    for my $file (@config_src) {
+        no warnings;
+        if ( !-f $file ) {
+            print "No pre-existing installed file found at $file\n";
+            next;
+        }
+        if ( open my $CONFIG, '-|', "\"$file\" --show-config" ) {
+            while (<$CONFIG>) {
+                if (/^([^\s=]+)=(.*)/) { $config{$1} = $2 }
+            }
+            close($CONFIG);
+        }
+        last if %config;
+    }
+    return %config;
+}
+
+### Handlers for configuration keys.
+# A handler takes a value assigned to a key and must return same or another
+# value which is considered valid for this key. I.e. it could set to a default
+# if the assigned value is undefined, for example.
+
+sub _ckey_libdir {
+    my ( $self, $val ) = @_;
+    unless ( defined $val ) {
+        return File::Spec->catdir( $self->{config}{prefix}, 'share' );
+    }
+    say "libdir: $val";
+    $val;
+}
+
+### Tieing-related stuff
+sub TIEHASH {
+    my $class = shift;
+    my $self  = bless {}, $class;
+    return $self->init(@_);
+}
+
+sub FETCH {
+    my $self = shift;
+    my $key  = shift;
+    return $self->{config}{$key};
+}
+
+sub STORE {
+    my $self = shift;
+    my ( $key, $val ) = @_;
+    my $mname = "_ckey_$key";
+    if ( $self->can($mname) ) {
+        $val = $self->$mname($val);
+    }
+    return $self->{config}{$key} = $val;
+}
+
+sub DELETE {
+    my $self = shift;
+    my $key  = shift;
+    delete $self->{config}{$key};
+}
+
+sub CLEAR {
+    my $self = shift;
+    $self->{config} = {};
+}
+
+sub EXISTS {
+    my $self = shift;
+    my $key  = shift;
+    exists $self->{config}{$key};
+}
+
+sub FIRSTKEY {
+    my $self = shift;
```

Comment:
`FIRSTKEY` should probably reset the internal iterator by adding this line before the call to `each …`

```perl
keys %{ $self->{config} };
```

(`keys` iterates the whole list so the iterator ends up reset after this call.)

Reply:
Not anymore. Tie support has been removed in the final version.